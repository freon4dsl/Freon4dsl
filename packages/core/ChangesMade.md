# Changes made to core and core-svelte

On 19-08-22:
1. Created README in /editor
2. Name of PiActions changed into PiCombinedActions, to distinguish this interface from 
/editor/actions/PiAction. Caused changes in meta, and changes in CustomActions in samples and test.
3. Everything todo with triggers moved from PiActions/PiCombinedActions and PiAction to new file 
/editor/actions/PiTrigger.
4. Moved utils that are specific for the editor from /util to /editor/util, and extracted PiCaret class
from BehaviorUtils.ts. Removed implementation of setCaret in TextBox, because this is always overwritten.
5. Replaced use of 'keyCode' by 'key', only the test whether a character is printable remains to be done.
6. General cleanup of code.
7. Changed 'keyPressAction' to 'isCharAllowed' in TextBox.
8. Cleanup of PiEditor.
9. Merged in the temp changes in tree balancing. 
