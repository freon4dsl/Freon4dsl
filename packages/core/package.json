{
  "name": "@freon4dsl/core",
  "version": "0.7.0-beta2",
  "description": "The uttermost core part of the Freon library, which enables you to build editors for the web",
  "publishConfig": {
    "access": "public"
  },
  "license": "MIT",
  "main": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "type": "module",
  "repository": {
    "type": "git",
    "url": "https://github.com/freon4dsl/Freon4dsl"
  },
  "files": [
    "/dist"
  ],
  "scripts": {
    "build": "rollup -c",
    "build-dev": "rollup -c",
    "clean": "rm -rf node_modules/ && rm -rf dist/",
    "build-release": "npm run build",
    "cleanup": "bash ../../scripts/cleanup.sh",
    "test": "vitest run",
    "start": "echo Start core",
    "release": "npm publish",
    "publish-local": "npm publish --registry http://localhost:4873",
    "unpublish-local": "npm unpublish --force --registry http://localhost:4873",
    "republish-local": "npm run unpublish-local && npm run publish-local"
  },
  "devDependencies": {
    "vitest": "^2.1.0",
    "@vitest/ui": "^2.1.0"
  },
  "dependencies": {
    "lodash": "^4.17.21",
    "@lionweb/validation": "0.6.2",
    "@lionweb/repository-client": "0.2.2",
    "mobx": "^6.12.3",
    "reflect-metadata": "^0.2.2",
    "tslib": "^2.6.3"
  }
}
