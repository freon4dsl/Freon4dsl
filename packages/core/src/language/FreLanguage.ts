import { ReferenceShortcut } from "../editor";
import { FreNode, FreModel, FreModelUnit } from "../ast";
import { EmptyStdLib, FreStdlib } from "../stdlib/index";
import { isNullOrUndefined } from "../util";
// import { FreLogger } from "../logging";
// const LOGGER = new FreLogger("Language");

export type PropertyKind = "primitive" | "part" | "reference";

export type PrimType = string | boolean | number;

export type FreLanguageProperty = {
    name: string;
    type: string;
    id?: string;
    key?: string; // used for LionWeb
    isList: boolean;
    isPublic: boolean;
    language: string;
    propertyKind: PropertyKind;
};
export type FreLanguageModel = {
    typeName: string;
    id?: string;
    key?: string; // used for LionWeb
    isNamespace?: boolean;
    isNamedElement?: boolean;
    isAbstract?: boolean;
    language: string;
    subConceptNames?: string[];
    properties: Map<string, FreLanguageProperty>;
    constructor: (id?: string) => FreModel;
    creator: (data: Partial<FreModel>) => FreModel;
    referenceShortcut?: ReferenceShortcut;
};
export type FreLanguageModelUnit = {
    typeName: string;
    id?: string;
    key?: string; // used for LionWeb
    // isPublic?: boolean;
    isNamespace?: boolean;
    isNamedElement?: boolean;
    isAbstract?: boolean;
    language: string;
    subConceptNames: string[];
    fileExtension: string;
    properties: Map<string, FreLanguageProperty>;
    constructor: (id?: string) => FreModelUnit;
    creator: (data: Partial<FreModelUnit>) => FreModelUnit;
    trigger: string;
    referenceShortcut?: ReferenceShortcut;
};
export type FreLanguageConcept = {
    typeName: string;
    id?: string;
    key?: string; // used for LionWeb
    isAbstract: boolean;
    isPublic: boolean;
    isLimited: boolean;
    instanceNames: string[];
    isNamespace?: boolean;
    isNamedElement?: boolean;
    language: string;
    baseName: string;
    subConceptNames: string[];
    properties: Map<string, FreLanguageProperty>;
    constructor: (id?: string) => FreNode;
    creator: (data: Partial<FreNode>) => FreNode;
    // Used by editor, therefore only in Concept
    trigger: string;
    referenceShortcut?: ReferenceShortcut;
};

export type FreLanguageInterface = {
    typeName: string;
    id?: string;
    key?: string; // used for LionWeb
    isPublic: boolean;
    isNamespace?: boolean;
    isNamedElement?: boolean;
    isAbstract?: boolean;
    subConceptNames: string[];
    properties: Map<string, FreLanguageProperty>;
    constructor?: (id?: string) => FreNode | undefined;
    creator?: (data: Partial<FreNode>) => FreNode | undefined;
    language: string;
    referenceShortcut?: ReferenceShortcut;
};

export type FreLanguageClassifier = FreLanguageModel | FreLanguageModelUnit | FreLanguageConcept | FreLanguageInterface;

export class FreLanguage {
    private static theInstance: FreLanguage;

    static getInstance() {
        if (FreLanguage.theInstance === undefined) {
            FreLanguage.theInstance = new FreLanguage();
        }
        return FreLanguage.theInstance;
    }

    private languageName: string;
    private languageId?: string;
    private pmodel: FreLanguageModel;
    private units: Map<string, FreLanguageModelUnit> = new Map<string, FreLanguageModelUnit>();
    private concepts: Map<string, FreLanguageConcept> = new Map<string, FreLanguageConcept>();
    private interfaces: Map<string, FreLanguageInterface> = new Map<string, FreLanguageInterface>();
    private _stdLib: FreStdlib = new EmptyStdLib();
    set stdLib(lib: FreStdlib) {
        this._stdLib = lib;
    }
    get stdLib(): FreStdlib {
        return this._stdLib;
    }

    private constructor() {}

    model(): FreLanguageModel {
        return this.pmodel;
    }

    modelOfType(typeName: string) {
        if (!!this.pmodel && this.pmodel.typeName === typeName) {
            return this.pmodel;
        } else {
            return null;
        }
    }

    unit(typeName: string): FreLanguageModelUnit | undefined {
        return this.units.get(typeName);
    }

    unitByKey(key: string): FreLanguageModelUnit | undefined {
        return this.helperByKey(this.units, key) as FreLanguageModelUnit;
    }

    concept(typeName: string): FreLanguageConcept | undefined {
        // console.log("Language find concept " + typeName);
        return this.concepts.get(typeName);
    }

    conceptByKey(conceptKey: string): FreLanguageConcept | undefined {
        return this.helperByKey(this.concepts, conceptKey) as FreLanguageConcept;
    }

    private helperByKey(
        map: Map<string, FreLanguageClassifier>,
        conceptKey: string,
    ): FreLanguageClassifier | undefined {
        for (const concept of map.values()) {
            if (concept.key === conceptKey) {
                return concept;
            }
        }
        return null;
    }

    interface(typeName: string): FreLanguageInterface | undefined {
        return this.interfaces.get(typeName);
    }

    interfaceByKey(key: string): FreLanguageInterface | undefined {
        return this.helperByKey(this.interfaces, key) as FreLanguageInterface;
    }

    classifier(typeName: string): FreLanguageClassifier | undefined {
        const concept1 = this.concepts.get(typeName);
        if (!!concept1) {
            return concept1;
        } else {
            const intf = this.interfaces.get(typeName);
            if (!!intf) {
                return intf;
            } else {
                const unit1 = this.units.get(typeName);
                if (!!unit1) {
                    return unit1;
                } else {
                    const model = this.modelOfType(typeName);
                    if (!!model) {
                        return model;
                    }
                }
            }
        }
        // console.log("RETURNING NULL FOR " + typeName)
        return undefined;
    }

    classifierByKey(key: string): FreLanguageClassifier | undefined {
        const concept1 = this.conceptByKey(key);
        if (!!concept1) {
            return concept1;
        } else {
            const intf = this.interfaceByKey(key);
            if (!!intf) {
                return intf;
            } else {
                const unit1 = this.unitByKey(key);
                if (!!unit1) {
                    return unit1;
                } else {
                    // TODO By Id for models
                    const model = this.modelOfType(key);
                    if (!!model) {
                        return model;
                    }
                }
            }
        }
        // console.log("RETURNING NULL FOR " + typeName)
        return undefined;
    }

    // private conceptProperty(typeName: string, propertyName: string): FreLanguageProperty | undefined {
    //     return this.concepts.get(typeName)?.properties.get(propertyName);
    // }
    //
    // private conceptPropertyByKey(conceptKey: string, propertyKey: string): FreLanguageProperty | undefined {
    //     return this.helperPropByKey(this.conceptByKey(conceptKey).properties, propertyKey);
    // }

    helperPropByKey(map: Map<string, FreLanguageProperty>, key: string): FreLanguageProperty | undefined {
        for (const prop of map.values()) {
            if (prop.key === key) {
                return prop;
            }
        }
        return undefined;
    }

    // private unitProperty(typeName: string, propertyName: string): FreLanguageProperty | undefined {
    //     return this.units.get(typeName)?.properties.get(propertyName);
    // }

    unitPropertyByKey(unitKey: string, propertyKey: string): FreLanguageProperty | undefined {
        // LOGGER.log("copnceptProperty [" + typeName + "."  + propertyName + "]");
        return this.helperPropByKey(this.unitByKey(unitKey).properties, propertyKey);
    }

    // private interfaceProperty(typeName: string, propertyName: string): FreLanguageProperty | undefined {
    //     return this.interfaces.get(typeName)?.properties.get(propertyName);
    // }

    interfacePropertyByKey(interfaceKey: string, propertyKey: string): FreLanguageProperty | undefined {
        // LOGGER.log("copnceptProperty [" + typeName + "."  + propertyName + "]");
        return this.helperPropByKey(this.interfaceByKey(interfaceKey).properties, propertyKey);
    }

    classifierProperty(typeName: string, propertyName: string): FreLanguageProperty | undefined {
        // LOGGER.log("CLASSIFIERPROPERTY " + typeName + "." + propertyName);
        const concept1 = this.concepts.get(typeName);
        if (!!concept1) {
            return concept1.properties.get(propertyName);
        } else {
            const intf = this.interfaces.get(typeName);
            if (!!intf) {
                return intf.properties.get(propertyName);
            } else {
                const unit1 = this.units.get(typeName);
                if (!!unit1) {
                    return unit1.properties.get(propertyName);
                } else {
                    const model = this.modelOfType(typeName);
                    if (!!model) {
                        return model.properties.get(propertyName);
                    }
                }
            }
        }
        return undefined;
    }

    classifierPropertyByKey(classifierKey: string, propertyKey: string): FreLanguageProperty | undefined {
        // LOGGER.log("CLASSIFIERPROPERTY " + typeName + "." + propertyName);
        const concept1 = this.classifierByKey(classifierKey);
        return this.helperPropByKey(concept1.properties, propertyKey);
    }

    allConceptProperties(typeName: string): FreLanguageProperty[] | undefined {
        // console.log("Looking up properties for "+ typeName);
        let myType: FreLanguageConcept | FreLanguageModelUnit | undefined = this.concept(typeName);
        if (isNullOrUndefined(myType)) {
            myType = this.unit(typeName);
        }
        if (myType === undefined) {
            return [];
        }
        return [...myType.properties.values()];
    }

    /**
     * Get all properties of kind "kind" of classifier "typename"
     * @param typename
     * @param ptype
     */
    public getPropertiesOfKind(typename: string, ptype: PropertyKind): FreLanguageProperty[] {
        const classifier: FreLanguageClassifier | undefined = FreLanguage.getInstance().classifier(typename);
        const foundProperties: FreLanguageProperty[] = [];
        if (!!classifier) {
            for (const prop of classifier.properties.values()) {
                if (prop.propertyKind === ptype) {
                    foundProperties.push(prop);
                }
            }
        }
        return foundProperties;
    }

    /**
     * Return the value of `prop' in `element'.
     * For ease of use, always returns a list, even is the property is not a list.
     * @param element
     * @param prop
     */
    public getPropertyValue(element: FreNode, prop: FreLanguageProperty): FreNode[] {
        if (prop.isList) {
            return element[prop.name];
        } else {
            const value = element[prop.name];
            if (!!value) {
                return [value];
            } else {
                return [];
            }
        }
    }

    /**
     * Return all named concept in the language.
     */
    public getNamedConcepts(): string[] {
        return Array.from(this.concepts.values())
            .filter((concept) => concept.isNamedElement)
            .map((concept) => concept.typeName);
    }

    /**
     * Return all named interfaces in the language.
     */
    public getNamedInterfaces(): string[] {
        return Array.from(this.interfaces.values())
            .filter((intfc) => intfc.isNamedElement)
            .map((intfc) => intfc.typeName);
    }

    /**
     * Return all named concepts and interfaces in the language.
     */
    public getNamedElements(): string[] {
        return this.getNamedConcepts().concat(this.getNamedInterfaces());
    }

    /**
     * Return the names of all model unit types.
     */
    public getUnitNames(): string[] {
        return Array.from(this.units.values()).map((unit) => unit.typeName);
    }

    createModel(id?: string): FreModel {
        return this.pmodel?.constructor(id);
    }

    createUnit(typeName: string, id?: string): FreModelUnit | undefined {
        return this.units.get(typeName)?.constructor(id);
    }

    /**
     * Create a new instance of the class `typeName`.
     * @param typeName
     */
    createConceptOrUnit(typeName: string, id?: string): FreNode | undefined {
        let myType: FreLanguageConcept | FreLanguageModelUnit | undefined = this.concept(typeName);
        if (isNullOrUndefined(myType)) {
            myType = this.unit(typeName);
        }
        return myType?.constructor(id);
    }

    addModel(model: FreLanguageModel) {
        if (!!this.pmodel) {
            console.error(
                "Language: adding model of type " +
                    model?.typeName +
                    " while there is already a model of type " +
                    this.pmodel.typeName,
            );
        }
        this.pmodel = model;
    }

    addUnit(unit: FreLanguageModelUnit) {
        this.units.set(unit.typeName, unit);
        unit.subConceptNames = [];
    }

    /**
     * Add a concept definition to this language
     * @param concept
     */
    addConcept(concept: FreLanguageConcept) {
        this.concepts.set(concept.typeName, concept);
    }

    addInterface(intface: FreLanguageInterface) {
        this.interfaces.set(intface.typeName, intface);
    }

    /**
     * Set Language name
     * @param name
     */
    set name(name: string) {
        this.languageName = name;
    }

    get name(): string {
        return this.languageName;
    }

    set id(id: string) {
        this.languageId = id;
    }

    get id(): string {
        return this.languageId;
    }

    subConcepts(typeName: string): string[] {
        const concept = this.concept(typeName);
        if (!!concept) {
            return concept.subConceptNames;
        }
        const intface = this.interface(typeName);
        if (!!intface) {
            return intface.subConceptNames;
        }
        return [];
    }

    referenceCreator: (name: string | string[], type: string) => any;

    addReferenceCreator(creator: (name: string | string[], type: string) => any) {
        this.referenceCreator = creator;
    }

    /**
     * Returns true if the freLanguageConcept of 'element', i.e. its metatype,
     * is the same as 'requestedType' or is a subtype of 'requestedType'.
     * @param element
     * @param requestedType
     */
    public metaConformsToType(element: FreNode, requestedType: string): boolean {
        const metatype = element.freLanguageConcept();
        return metatype === requestedType || FreLanguage.getInstance().subConcepts(requestedType).includes(metatype);
    }
}
