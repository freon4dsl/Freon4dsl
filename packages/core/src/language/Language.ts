import { ReferenceShortcut } from "../editor";
import { PiElement, PiModel, PiModelUnit } from "../ast";
import { isNullOrUndefined } from "../util";
import { PiLogger } from "../logging";
const LOGGER = new PiLogger("Language");

export type PropertyKind = "primitive" | "part" | "reference";

export type Property = {
    name: string;
    type: string;
    isList: boolean;
    isPublic: boolean;
    propertyKind: PropertyKind;
};
export type Model = {
    typeName: string;
    isNamespace?: boolean;
    isNamedElement?: boolean;
    subConceptNames?: string[];
    properties: Map<string, Property>;
    constructor: () => PiModel;
};
export type ModelUnit = {
    typeName: string;
    // isPublic?: boolean;
    isNamespace?: boolean;
    isNamedElement?: boolean;
    subConceptNames?: string[];
    fileExtension: string;
    properties: Map<string, Property>;
    constructor: () => PiModelUnit;
};
export type Concept = {
    typeName: string;
    isAbstract: boolean;
    isPublic: boolean;
    isNamespace?: boolean;
    isNamedElement?: boolean;
    baseName: string;
    subConceptNames: string[];
    properties: Map<string, Property>;
    constructor: () => PiElement;
    // Used by editor, therefore only in Concept
    trigger: string;
    referenceShortcut?: ReferenceShortcut;
};

export type Interface = {
    typeName: string;
    isPublic: boolean;
    isNamespace?: boolean;
    isNamedElement?: boolean;
    subConceptNames: string[];
    properties: Map<string, Property>;
};

export type Classifier = Model | ModelUnit | Concept | Interface;

export class Language {
    private static theInstance: Language = null;

    static getInstance() {
        if (Language.theInstance === null) {
            Language.theInstance = new Language();
        }
        return Language.theInstance;
    }

    private languageName: string;
    private pmodel: Model;
    private units: Map<string, ModelUnit> = new Map<string, ModelUnit>();
    private concepts: Map<string, Concept> = new Map<string, Concept>();
    private interfaces: Map<string, Interface> = new Map<string, Interface>();

    private constructor() {
    }

    model(): Model {
        return this.pmodel;
    }

    modelOfType(typeName: string) {
        if(!!this.pmodel && (this.pmodel.typeName === typeName)) {
            return this.pmodel;
        } else {
            return null;
        }
    }

    unit(typeName: string): ModelUnit {
        return this.units.get(typeName);
    }

    concept(typeName: string): Concept {
        // console.log("Language find concept " + typeName);
        return this.concepts.get(typeName);
    }

    interface(typeName: string): Interface {
        return this.interfaces.get(typeName);
    }

    classifier(typeName: string): Classifier {
        let concept1 = this.concepts.get(typeName);
        if (!!concept1) {
            return concept1;
        } else {
            let intf = this.interfaces.get(typeName);
            if (!!intf) {
                return intf;
            } else {
                let unit1 = this.units.get(typeName);
                if (!!unit1) {
                    return unit1;
                } else {
                    let model = this.modelOfType(typeName);
                    if(!! model){
                        return model;
                    }
                }
            }
        }
        console.log("RETURNING NULL FOR " + typeName)
        return null;
    }

    conceptProperty(typeName: string, propertyName: string): Property {
        // LOGGER.log("copnceptProperty [" + typeName + "."  + propertyName + "]");
        return this.concepts.get(typeName).properties.get(propertyName);
    }

    unitProperty(typeName: string, propertyName: string): Property {
        return this.units.get(typeName).properties.get(propertyName);
    }

    interfaceProperty(typeName: string, propertyName: string): Property {
        return this.interfaces.get(typeName).properties.get(propertyName);
    }

    classifierProperty(typeName: string, propertyName: string): Property {
        // LOGGER.log("CLASSIFIERPROPERTY " + typeName + "." + propertyName);
        let concept1 = this.concepts.get(typeName);
        if (!!concept1) {
            return concept1.properties.get(propertyName);
        } else {
            let intf = this.interfaces.get(typeName);
            if (!!intf) {
                return intf.properties.get(propertyName);
            } else {
                let unit1 = this.units.get(typeName);
                if (!!unit1) {
                    return unit1.properties.get(propertyName);
                }  else {
                    let model = this.modelOfType(typeName);
                    if( !!model){
                        return model.properties.get(propertyName);
                    }
                }
            }
        }
        return null;
    }

    allConceptProperties(typeName: string): IterableIterator<Property> {
        // console.log("Looking up properties for "+ typeName);
        let myType: Concept | ModelUnit = this.concept(typeName);
        if (isNullOrUndefined(myType)) {
            myType = this.unit(typeName);
        }
        return myType?.properties.values();
    }

    /**
     * Get all properties of kind "kind" of classifier "typename"
     * @param typename
     * @param ptype
     */
    public getPropertiesOfKind(typename: string, ptype: PropertyKind): Property[]  {
        let classifier: Classifier = Language.getInstance().classifier(typename);
        const foundProperties: Property[] = [];
        if (!!classifier) {
            for (const prop of classifier.properties.values()) {
                if (prop.propertyKind === ptype) {
                    foundProperties.push(prop);
                }
            }
        }
        return foundProperties;
    }

    /**
     * Return the value of `prop' in `element'.
     * For ease of use, always returns a list, even is the property is not a list.
     * @param element
     * @param prop
     */
    public getPropertyValue(element: PiElement, prop: Property): PiElement[] {
        if( prop.isList){
            return element[prop.name];
        } else {
            const value = element[prop.name];
            if( !!value) {
                return [value];
            } else {
                return [];
            }
        }
    }
    createModel(): PiModel {
        return this.pmodel?.constructor();
    }

    createUnit(typeName: string): PiModelUnit {
        return this.units.get(typeName).constructor();
    }

    /**
     * Create a new instance of the class `typeName`.
     * @param typeName
     */
    createConceptOrUnit(typeName: string): PiElement {
        let myType: Concept | ModelUnit = this.concept(typeName);
        if (isNullOrUndefined(myType)) {
            myType = this.unit(typeName);
        }
        return myType?.constructor();
    }


    addModel(model: Model) {
        if(!!this.pmodel) {
            console.error("Language: adding model of type " + model?.typeName + " while there is already a model of type " + this.pmodel.typeName);
        }
        this.pmodel = model;
    }

    addUnit(unit: ModelUnit) {
        this.units.set(unit.typeName, unit);
        unit.subConceptNames = [];
    }

    /**
     * Add a concept definition to this language
     * @param concept
     */
    addConcept(concept: Concept) {
        this.concepts.set(concept.typeName, concept);
    }

    addInterface(intface: Interface) {
        this.interfaces.set(intface.typeName, intface);
    }

    /**
     * Set Language name
     * @param name
     */
    set name(name: string) {
        this.languageName = name;
    }

    get name(): string {
        return this.languageName;
    }

    subConcepts(typeName: string): string[] {
        const concept = this.concept(typeName);
        if (!!concept) {
            return concept.subConceptNames;
        }
        const intface = this.interface(typeName);
        if (!!intface) {
            return intface.subConceptNames;
        }
        return [];
    }

    referenceCreator: (name: string, type: string) => any;

    addReferenceCreator(creator: (name: string, type: string) => any) {
        this.referenceCreator = creator;
    }

    /**
     * Returns true if the piLanguageConcept of 'element', i.e. its metatype,
     * is the same as 'requestedType' or is a subtype of 'requestedType'.
     * @param element
     * @param requestedType
     */
    public metaConformsToType(element: PiElement, requestedType: string): boolean {
        const metatype = element.piLanguageConcept();
        if (metatype === requestedType || Language.getInstance().subConcepts(requestedType).includes(metatype)) {
            return true;
        }
        return false;
    }
}
