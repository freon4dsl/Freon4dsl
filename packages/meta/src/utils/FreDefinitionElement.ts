import { ParseLocation, FreParseLocation } from "./parsingAndChecking/FreGenericParser";

// root of the inheritance structure of all elements that can be present in any of the definitions
export class FreDefinitionElement {
    location: ParseLocation;        // used to store the location information from the PEGJS parser
    agl_location: FreParseLocation;  // used to store the location information from the AGL parser
}
