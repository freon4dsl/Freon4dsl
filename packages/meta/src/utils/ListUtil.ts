export class ListUtil {
    static addIfNotPresent<T>(list: T[], addition: T) {
        if (!!addition && !list.includes(addition)) {
            list.push(addition);
        }
    }

    static addListIfNotPresent<T>(list: T[], additionsList: T[]) {
        additionsList.forEach(extra => {
            ListUtil.addIfNotPresent<T>(list, extra);
        })
    }
}
