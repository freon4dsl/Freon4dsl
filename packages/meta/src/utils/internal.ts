export * from "./PiDefinitionElement";
export * from "./ListUtil";
export * from "./UserLogger";
export * from "./MetaLogger";
export * from "./parsingAndChecking";
export * from "./generation";
