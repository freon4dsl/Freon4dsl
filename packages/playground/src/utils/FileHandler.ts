import * as fs from "fs";

export class FileHandler {

    /**
     * Reads a string from the file located at 'filepath'. If the
     * file is not present an Error will be thrown.
     * @param filepath
     */
    public stringFromFile(filepath: string): string {
        // read language file
        if (!fs.existsSync(filepath)) {
            console.error(this, "File '" + filepath + "' does not exist, exiting.");
            throw new Error(`File '${filepath}' not found.`);
        }
        return fs.readFileSync(filepath, { encoding: "utf8" });
    }

    /**
     * Writes a string to the file located at 'filepath'. If the
     * file is not present it will be created.
     * May throw an Error if the file cannot be written or created.
     * @param filepath
     * @param output
     */
    public stringToFile(filepath: string, output: string) {
        if (fs.existsSync(filepath)) {
            console.log(this, "FileHandler: file " + filepath + " already exists, overwriting it.");
        }
        fs.writeFileSync(filepath, output);
    }

}
