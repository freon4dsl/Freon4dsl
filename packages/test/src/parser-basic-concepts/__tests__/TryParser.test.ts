import { FreUtils } from "@freon4dsl/core";
import { FileHandler } from "../../utils/FileHandler";
import { TestConceptsModelEnvironment } from "../config/gen/TestConceptsModelEnvironment";
import { ExpressionTest, TestConceptsModel } from "../language/gen";
import { describe, test, expect, beforeEach } from "vitest";

describe("Parser concepts of type", () => {
    const reader = TestConceptsModelEnvironment.getInstance().reader;
    const writer = TestConceptsModelEnvironment.getInstance().writer;
    const fileHandler = new FileHandler();

    beforeEach(() => {
        // Ensure that ID's of created elements do not change.
        FreUtils.resetId();
    });

    test(" Expression ", () => {
        let unit1: ExpressionTest = undefined;
        try {
            const input = fileHandler.stringFromFile("src/parser-basic-concepts/__inputs__/test1.exp");
            unit1 = reader.readFromString(
                input,
                "ExpressionTest",
                new TestConceptsModel(),
                "src/parser-basic-concepts/__inputs__/test1.exp"
            ) as ExpressionTest;
        } catch (e) {
            // console.log(e.stack);
            expect(e).toBeNaN();
        }
        expect (unit1).not.toBeUndefined();
        // console.log(writer.writeToString(unit1, 0, false));
        expect(unit1).toMatchSnapshot();
    });
});
