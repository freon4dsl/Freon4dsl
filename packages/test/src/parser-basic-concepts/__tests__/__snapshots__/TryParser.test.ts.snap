// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Parser concepts of type >  Expression  1`] = `
ExpressionTest {
  "$$owner": TestConceptsModel {
    "$$owner": null,
    "$$propertyIndex": undefined,
    "$$propertyName": "",
    "$id": "ID-1",
    "$typename": "TestConceptsModel",
    "_FRE_expr": [Circular],
  },
  "$$propertyIndex": undefined,
  "$$propertyName": "expr",
  "$id": "ID-12",
  "$typename": "ExpressionTest",
  "_FRE_binEx": [
    PlusEx {
      "$$owner": [Circular],
      "$$propertyIndex": 0,
      "$$propertyName": "binEx",
      "$id": "ID-5",
      "$typename": "PlusEx",
      "_FRE_left": StringLiteralEx {
        "$$owner": [Circular],
        "$$propertyIndex": undefined,
        "$$propertyName": "left",
        "$id": "ID-2",
        "$typename": "StringLiteralEx",
        "_FRE_value": "literal1",
        "parseLocation": FreParseLocation {
          "column": 2,
          "filename": "src/parser-basic-concepts/__inputs__/test1.exp",
          "line": 4,
        },
      },
      "_FRE_right": StringLiteralEx {
        "$$owner": [Circular],
        "$$propertyIndex": undefined,
        "$$propertyName": "right",
        "$id": "ID-4",
        "$typename": "StringLiteralEx",
        "_FRE_value": "literal2",
        "parseLocation": FreParseLocation {
          "column": 39,
          "filename": "src/parser-basic-concepts/__inputs__/test1.exp",
          "line": 4,
        },
      },
      "parseLocation": FreParseLocation {
        "column": 2,
        "filename": "src/parser-basic-concepts/__inputs__/test1.exp",
        "line": 4,
      },
    },
    DivideEx {
      "$$owner": [Circular],
      "$$propertyIndex": 1,
      "$$propertyName": "binEx",
      "$id": "ID-11",
      "$typename": "DivideEx",
      "_FRE_left": NumberLiteralEx {
        "$$owner": [Circular],
        "$$propertyIndex": undefined,
        "$$propertyName": "left",
        "$id": "ID-6",
        "$typename": "NumberLiteralEx",
        "_FRE_value": 109,
        "parseLocation": FreParseLocation {
          "column": 2,
          "filename": "src/parser-basic-concepts/__inputs__/test1.exp",
          "line": 5,
        },
      },
      "_FRE_right": MultiplyEx {
        "$$owner": [Circular],
        "$$propertyIndex": undefined,
        "$$propertyName": "right",
        "$id": "ID-10",
        "$typename": "MultiplyEx",
        "_FRE_left": StringLiteralEx {
          "$$owner": [Circular],
          "$$propertyIndex": undefined,
          "$$propertyName": "left",
          "$id": "ID-7",
          "$typename": "StringLiteralEx",
          "_FRE_value": "literal3",
          "parseLocation": FreParseLocation {
            "column": 32,
            "filename": "src/parser-basic-concepts/__inputs__/test1.exp",
            "line": 5,
          },
        },
        "_FRE_right": BooleanLiteralEx {
          "$$owner": [Circular],
          "$$propertyIndex": undefined,
          "$$propertyName": "right",
          "$id": "ID-9",
          "$typename": "BooleanLiteralEx",
          "_FRE_value": true,
          "parseLocation": FreParseLocation {
            "column": 69,
            "filename": "src/parser-basic-concepts/__inputs__/test1.exp",
            "line": 5,
          },
        },
        "parseLocation": FreParseLocation {
          "column": 32,
          "filename": "src/parser-basic-concepts/__inputs__/test1.exp",
          "line": 5,
        },
      },
      "parseLocation": FreParseLocation {
        "column": 2,
        "filename": "src/parser-basic-concepts/__inputs__/test1.exp",
        "line": 5,
      },
    },
  ],
  "_FRE_name": "simpleExpressions",
  "_FRE_normalEx": [],
  "fileExtension": "",
  "parseLocation": FreParseLocation {
    "column": 1,
    "filename": "src/parser-basic-concepts/__inputs__/test1.exp",
    "line": 1,
  },
}
`;
